lib/error.js

```diff
@@ -2,9 +2,47 @@ var makeError = require('make-error');
 
 function ShareDBError(code, message) {
   ShareDBError.super.call(this, message);
-  this.code = code;
+  this.code = code || ShareDBError.code.UNKNOWN_ERROR;
 }
 
 makeError(ShareDBError);
 
+ShareDBError.code = {
+  BADLY_FORMED_MESSAGE: 'ERR_BADLY_FORMED_MESSAGE',
+  CANNOT_PROJECT_TYPE: 'ERR_CANNOT_PROJECT_TYPE',
+  CREATE_DATA_MUST_BE_AN_OBJECT: 'ERR_CREATE_DATA_MUST_BE_AN_OBJECT',
+  DATABASE_ADAPTER_NOT_FOUND: 'ERR_DATABASE_ADAPTER_NOT_FOUND',
+  DEL_MUST_BE_TRUE: 'ERR_DEL_MUST_BE_TRUE',
+  DOCUMENT_ALREADY_CREATED: 'ERR_DOCUMENT_ALREADY_CREATED',
+  DOCUMENT_CREATED_REMOTELY: 'ERR_DOCUMENT_CREATED_REMOTELY',
+  DOCUMENT_DOES_NOT_EXIST: 'ERR_DOCUMENT_DOES_NOT_EXIST',
+  DOCUMENT_WAS_DELETED: 'ERR_DOCUMENT_WAS_DELETED',
+  INVALID_CLIENT_ID: 'ERR_INVALID_CLIENT_ID',
+  INVALID_DEFAULT_TYPE: 'ERR_INVALID_DEFAULT_TYPE',
+  INVALID_METADATA: 'ERR_INVALID_METADATA',
+  INVALID_SRC: 'ERR_INVALID_SRC',
+  INVALID_SEQ: 'ERR_INVALID_SEQ',
+  INVALID_STATE_TRANSITION: 'ERR_INVALID_STATE_TRANSITION',
+  INVALID_PROTOCOL_VERSION: 'ERR_INVALID_PROTOCOL_VERSION',
+  INVALID_VERSION: 'ERR_INVALID_VERSION',
+  LISTENER_NOT_ASSIGNED: 'ERR_LISTENER_NOT_ASSIGNED',
+  MAX_RETRIES: 'ERR_MAX_RETRIES',
+  MISSING_CREATE_TYPE: 'ERR_MISSING_CREATE_TYPE',
+  MISSING_OP: 'ERR_MISSING_OP',
+  MISSING_TRANSFORM_OPS: 'ERR_MISSING_TRANSFORM_OPS',
+  MISSING_SNAPSHOT: 'ERR_MISSING_SNAPSHOT',
+  MISSING_SNAPSHOT_VERSION: 'ERR_MISSING_SNAPSHOT_VERSION',
+  NO_OP_TO_SEND: 'ERR_NO_OP_TO_SEND',
+  NOT_IMPLEMENTED: 'ERR_NOT_IMPLEMENTED',
+  OP_ALREADY_SUBMITTED: 'ERR_OP_ALREADY_SUBMITTED',
+  OP_INVALID_IN_PROJECTION: 'ERR_OP_INVALID_IN_PROJECTION',
+  OP_SUBMIT_REJECTED: 'ERR_OP_SUBMIT_REJECTED',
+  OP_VERSION_NEWER_THAN_SNAPSHOT: 'ERR_OP_VERSION_NEWER_THAN_SNAPSHOT',
+  SRC_AND_SEQ_MUST_BE_SET_TOGETHER: 'ERR_SRC_AND_SEQ_MUST_BE_SET_TOGETHER',
+  SUBSCRIBE_UNSUPPORTED: 'ERR_SUBSCRIBE_UNSUPPORTED',
+  UNKNOWN_ERROR: 'ERR_UNKNOWN_ERROR',
+  UNKNOWN_TYPE: 'ERR_UNKNOWN_TYPE',
+  VERSION_MISMATCH: 'ERR_VERSION_MISMATCH'
```

Comment:
-> `OP_SNAPSHOT_VERSION_MISMATCH`?

Reply:
Also all the different version mismatches should be different error codes, to help you narrow down where the error came from.